backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/CodeGenerator.kt

```diff
@@ -18,6 +18,39 @@ import org.jetbrains.kotlin.ir.util.*
 import org.jetbrains.kotlin.backend.konan.descriptors.resolveFakeOverride
 import org.jetbrains.kotlin.backend.konan.ir.*
 import org.jetbrains.kotlin.descriptors.konan.CompiledKlibModuleOrigin
+import org.jetbrains.kotlin.ir.expressions.IrConst
+import org.jetbrains.kotlin.ir.expressions.impl.IrDelegatingConstructorCallImpl
+import org.jetbrains.kotlin.ir.expressions.impl.IrReturnImpl
+import org.jetbrains.kotlin.ir.types.isPrimitiveType
+import org.jetbrains.kotlin.ir.types.isString
+
+enum class ObjectStorageKind {
+    PERMANENT,
+    THREAD_LOCAL,
+    SHARED
+}
+
+private val IrConstructor.isAnyConstructorDelegation: Boolean
+    get() {
+        val statements = this.body?.statements ?: return false
+        if (statements.size != 2) return false
+        if (statements[1] !is IrReturnImpl) return false
+        val constructorCall = statements[0] as? IrDelegatingConstructorCallImpl ?: return false
+        val constructor = constructorCall.symbol.owner as? IrConstructor ?: return false
+        if (!constructor.constructedClass.isAny()) return false
+        return true
+    }
+
+// TODO: shall we memoize that property?
+internal fun IrClass.hasConstStateAndNoSideEffects(context: Context): Boolean {
+    val fields = context.getLayoutBuilder(this).fields
+    if (this.hasAnnotation(KonanFqNames.canBePrecreated)) {
+        assert(fields.all { it.isFinal && (it.type.isPrimitiveType() || it.type.isString()) &&
+                it.initializer?.expression is IrConst<*> })
+        return true
+    }
+    return fields.isEmpty() && this.constructors.all { it.isAnyConstructorDelegation }
```

Comment:
What if this class is cached and its IR is just a stub?

Reply:
Why there's a difference?

Reply:
External IR stubs may omit fields as non-public interface.